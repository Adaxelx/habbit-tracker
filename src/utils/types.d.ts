export type GridProps = {
  events: Event[];
  month: number;
  year: number;
  moveDate: Function;
};

export interface Event {
  _id: string;
  daysOfWeek: number[];
  title: string;
  timeStart?: string;
  timeEnd?: string;
  dateStart: string;
  dateEnd: string;
  label?: string | Label;
  description?: string;
  userId: string;
  checked?: boolean;
}

export interface EventLoop extends Event {
  numericStart: number;
  numericEnd: number;
}

export type Label = {
  _id: string;
  title: string;
  color: string;
};

export type CalendarTile = {
  day: number;
  events: EventLoop[];
  id: number;
};

export enum SIDES {
  LEFT = -1,
  RIGHT = 1,
}
