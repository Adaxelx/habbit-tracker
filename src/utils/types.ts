export type GridProps = {
  events: Event[];
  month: number;
  year: number;
  moveDate: Function;
  handleDayChange: Function;
};

export interface Event {
  _id: string;
  daysOfWeek: number[];
  title: string;
  timeStart?: string;
  timeEnd?: string;
  dateStart: string;
  dateEnd: string;
  label?: string | Label;
  description?: string;
  userId: string;
  checked?: [
    {
      day: number;
      month: number;
      year: number;
      _id: string;
    },
  ];
}

export interface EventLoop extends Event {
  numericStart: number;
  numericEnd: number;
}

export type Label = {
  _id: string;
  title: string;
  color: string;
  userId: string;
};

export type CalendarTile = {
  day: number;
  events: EventLoop[];
  id: number;
};

export enum SIDES {
  LEFT = -1,
  RIGHT = 1,
}

export type Color = { color?: string; id?: string };

export type NavProps = {
  header: string;
  moveDate: Function;
};

export interface DayCardProps {
  day: string;
  token: string | undefined;
}
