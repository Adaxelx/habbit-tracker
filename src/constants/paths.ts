const path = {
  DASHBOARD: '/',
  CALENDAR: '/calendar',
  LOGIN: '/login',
  REGISTER: '/register',
};

export default path;
