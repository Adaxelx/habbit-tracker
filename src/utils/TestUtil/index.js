export { default } from './TestUtil.ts';
