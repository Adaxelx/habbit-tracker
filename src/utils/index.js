export { default as TestUtil } from './TestUtil';
export { default as RenderWithRouter } from './RenderWithRouter';
export * from './validators';
export * from './types.d';
