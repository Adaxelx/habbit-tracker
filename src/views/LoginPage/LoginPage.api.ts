import APIpaths from 'constants/APIpaths';

export type LoginInputs = {
  login: string;
  password: string;
};

export const loginUser = async (data: LoginInputs) => {
  const response = await fetch(APIpaths.LOGIN, { method: 'POST', body: JSON.stringify(data) });
  if (response.status === 200) {
    return response.json();
  }
  throw new Error('Something went wrong.');
};
