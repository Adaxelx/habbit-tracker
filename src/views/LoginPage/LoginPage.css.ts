import styled from 'styled-components';
import { FlexColCenter } from 'styles';

export const StyledWrapper = styled(FlexColCenter)`
  height: 100vh;
  width: 100%;
  &::last-child {
    margin-bottom: 0;
  }
`;
