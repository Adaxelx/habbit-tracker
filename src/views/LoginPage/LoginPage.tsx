import React, { useState } from 'react';
import { Input, Container, Button, Alert } from 'components';
import { useForm } from 'react-hook-form';
import { useHistory, Redirect } from 'react-router-dom';
import paths from 'constants/paths';
import { useUserContext } from 'context';
import { createRestrictedLengthObject } from 'utils';
import { StyledWrapper } from './LoginPage.css';
import { loginUser, LoginInputs } from './LoginPage.api';

const LoginPage = () => {
  const [error, setError] = useState('');
  const [loading, setLoading] = useState(false);
  const user = useUserContext();

  const { register, handleSubmit, errors } = useForm<LoginInputs>();
  const history = useHistory();

  const onSubmit = async (data: LoginInputs) => {
    setLoading(true);
    try {
      const response = await loginUser(data);
      setLoading(false);
      user.login(response.token);
      history.push(paths.CALENDAR);
    } catch (err) {
      setError(err.message);
      setLoading(false);
    }
  };

  return user.token ? (
    <Redirect to={paths.CALENDAR} />
  ) : (
    <Container>
      <StyledWrapper as="form" onSubmit={handleSubmit(onSubmit)} noValidate>
        <Input
          name="login"
          label="Login"
          id="login"
          refVal={register(createRestrictedLengthObject('login'))}
          minLength={6}
          maxLength={30}
          error={errors.login}
          data-testid="login"
        />
        <Input
          name="password"
          label="Password"
          id="password"
          type="password"
          refVal={register(createRestrictedLengthObject('password'))}
          autoComplete="on"
          minLength={6}
          maxLength={30}
          error={errors.password}
          data-testid="password"
        />
        <Alert error={error} loading={loading} />
        <Button size="m" type="submit" data-testid="submit">
          Log in
        </Button>
      </StyledWrapper>
    </Container>
  );
};

export default LoginPage;
