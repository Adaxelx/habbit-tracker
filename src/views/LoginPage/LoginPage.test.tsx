import React from 'react';
import 'jest-styled-components';
import { waitFor } from '@testing-library/react';
import paths from 'constants/paths';
import { TestUtil } from 'utils';
import LoginPage from './LoginPage';

describe('LoginPage', () => {
  let util: TestUtil;
  let fail: boolean;

  const mockedFetch = (input: string, init: { method: string }) => {
    if (fail) {
      return Promise.resolve({
        status: 500,
        ok: false,
        json: () => Promise.resolve({ message: 'Something went wrong.' }),
      });
    }
    switch (init.method) {
      case 'POST':
        return Promise.resolve({
          status: 200,
          ok: true,
          json: () => Promise.resolve({ token: 'cos' }),
        });

      default:
        return null;
    }
  };

  beforeAll(() => {
    jest.spyOn(window, 'fetch');
  });
  beforeEach(() => {
    jest.clearAllMocks();
    window.fetch.mockImplementation(mockedFetch);
    util = new TestUtil(<LoginPage />);
    fail = false;
  });

  it('should match snapshot', () => {
    expect(util.render.asFragment()).toMatchSnapshot();
  });

  it('should send login data', async () => {
    util.setValue('login', 'login1234');
    util.setValue('password', 'password123');
    util.click('submit');
    await waitFor(() => expect(fetch).toHaveBeenCalled());
    await waitFor(() => expect(util.render.history.location.pathname).toEqual(paths.CALENDAR));
  });

  it('should log alert on error', async () => {
    fail = true;
    util.setValue('login', 'login1234');
    util.setValue('password', 'password123');
    util.click('submit');
    await waitFor(() => expect(fetch).toHaveBeenCalled());
    expect(util.get('alert').textContent).toBe('Something went wrong.');
  });

  it('should not call fetch if validation fails', async () => {
    util.setValue('login', 'log');
    util.click('submit');
    await waitFor(() => expect(fetch).not.toHaveBeenCalled());
  });
});
