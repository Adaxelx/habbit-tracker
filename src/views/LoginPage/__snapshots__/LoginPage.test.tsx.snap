// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage should match snapshot 1`] = `
<DocumentFragment>
  .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 {
  height: 100vh;
  width: 100%;
}

.c2::last-child {
  margin-bottom: 0;
}

.c6 {
  color: #fff;
  background-color: #FFC136;
  border: 1px solid #707070;
  padding: 1.25rem;
  font-size: 1.5rem;
  min-width: 150px;
  display: inline-block;
  text-align: center;
  margin-top: 16px;
  margin-left: 0;
  margin-right: 0;
}

.c6:hover {
  cursor: pointer;
}

.c0 {
  min-height: 100vh;
  width: 100%;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 1rem;
  width: 100%;
  max-width: 300px;
}

.c4 {
  font-size: 1.5rem;
  margin-bottom: 0.75rem;
}

.c5 {
  font-size: 1.25rem;
  width: 100%;
  max-width: 300px;
}

.c5:focus {
  outline: thin dotted;
}

<main
    class="c0"
  >
    <form
      class="c1 c2"
      novalidate=""
    >
      <div
        class="c3"
      >
        <label
          class="c4"
          for="login"
        >
          Login
        </label>
        <input
          class="c5"
          data-testid="login"
          id="login"
          maxlength="30"
          minlength="6"
          name="login"
          type="text"
        />
      </div>
      <div
        class="c3"
      >
        <label
          class="c4"
          for="password"
        >
          Password
        </label>
        <input
          autocomplete="on"
          class="c5"
          data-testid="password"
          id="password"
          maxlength="30"
          minlength="6"
          name="password"
          type="password"
        />
      </div>
      <button
        class="c6"
        data-testid="submit"
        type="submit"
      >
        Log in
      </button>
    </form>
  </main>
</DocumentFragment>
`;
